"use strict";(self.webpackChunkmissing_semester=self.webpackChunkmissing_semester||[]).push([[465],{3905:(a,e,n)=>{n.d(e,{Zo:()=>u,kt:()=>c});var r=n(7294);function t(a,e,n){return e in a?Object.defineProperty(a,e,{value:n,enumerable:!0,configurable:!0,writable:!0}):a[e]=n,a}function i(a,e){var n=Object.keys(a);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(a);e&&(r=r.filter((function(e){return Object.getOwnPropertyDescriptor(a,e).enumerable}))),n.push.apply(n,r)}return n}function o(a){for(var e=1;e<arguments.length;e++){var n=null!=arguments[e]?arguments[e]:{};e%2?i(Object(n),!0).forEach((function(e){t(a,e,n[e])})):Object.getOwnPropertyDescriptors?Object.defineProperties(a,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(e){Object.defineProperty(a,e,Object.getOwnPropertyDescriptor(n,e))}))}return a}function l(a,e){if(null==a)return{};var n,r,t=function(a,e){if(null==a)return{};var n,r,t={},i=Object.keys(a);for(r=0;r<i.length;r++)n=i[r],e.indexOf(n)>=0||(t[n]=a[n]);return t}(a,e);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(a);for(r=0;r<i.length;r++)n=i[r],e.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(a,n)&&(t[n]=a[n])}return t}var s=r.createContext({}),p=function(a){var e=r.useContext(s),n=e;return a&&(n="function"==typeof a?a(e):o(o({},e),a)),n},u=function(a){var e=p(a.components);return r.createElement(s.Provider,{value:e},a.children)},k="mdxType",m={inlineCode:"code",wrapper:function(a){var e=a.children;return r.createElement(r.Fragment,{},e)}},d=r.forwardRef((function(a,e){var n=a.components,t=a.mdxType,i=a.originalType,s=a.parentName,u=l(a,["components","mdxType","originalType","parentName"]),k=p(n),d=t,c=k["".concat(s,".").concat(d)]||k[d]||m[d]||i;return n?r.createElement(c,o(o({ref:e},u),{},{components:n})):r.createElement(c,o({ref:e},u))}));function c(a,e){var n=arguments,t=e&&e.mdxType;if("string"==typeof a||t){var i=n.length,o=new Array(i);o[0]=d;var l={};for(var s in e)hasOwnProperty.call(e,s)&&(l[s]=e[s]);l.originalType=a,l[k]="string"==typeof a?a:t,o[1]=l;for(var p=2;p<i;p++)o[p]=n[p];return r.createElement.apply(null,o)}return r.createElement.apply(null,n)}d.displayName="MDXCreateElement"},2663:(a,e,n)=>{n.r(e),n.d(e,{assets:()=>s,contentTitle:()=>o,default:()=>k,frontMatter:()=>i,metadata:()=>l,toc:()=>p});var r=n(7462),t=(n(7294),n(3905));const i={sidebar_position:6,title:"5. Mengenali Error"},o="Error",l={unversionedId:"error",id:"error",title:"5. Mengenali Error",description:"xkcd 2200: Unreachable State",source:"@site/docs/error.mdx",sourceDirName:".",slug:"/error",permalink:"/error",draft:!1,editUrl:"https://github.com/kambi-ng/missing-semester/edit/main/docs/error.mdx",tags:[],version:"current",sidebarPosition:6,frontMatter:{sidebar_position:6,title:"5. Mengenali Error"},sidebar:"tutorialSidebar",previous:{title:"4. Git",permalink:"/git"}},s={},p=[{value:"Mengidentifikasi Error",id:"mengidentifikasi-error",level:2},{value:"Membaca Stack Trace",id:"membaca-stack-trace",level:2}],u={toc:p};function k(a){let{components:e,...n}=a;return(0,t.kt)("wrapper",(0,r.Z)({},u,n,{components:e,mdxType:"MDXLayout"}),(0,t.kt)("h1",{id:"error"},"Error"),(0,t.kt)("center",null,(0,t.kt)("img",{src:"https://imgs.xkcd.com/comics/unreachable_state.png"})),(0,t.kt)("center",null,(0,t.kt)("a",{href:"https://xkcd.com/2200"},"xkcd 2200: Unreachable State")),(0,t.kt)("p",null,"Salah satu hal yang sering ditemui selama menjadi mahasiswa Ilmu Komputer (atau bahkan cukup menggunakan komputer) adalah error. Error dapat terjadi karena banyak hal, namun pada umumnya terjadi karena terdapat suatu hal yang tidak terduga atau diinginkan. Hal ini dapat terjadi dalam program yang biasa kita gunakan, atau saat kita melakukan pemrograman. Topik ini akan menginformasikan kamu bagaimana cara merespons error-error tersebut."),(0,t.kt)("h2",{id:"mengidentifikasi-error"},"Mengidentifikasi Error"),(0,t.kt)("center",null,(0,t.kt)("img",{src:"https://imgs.xkcd.com/comics/error_code.png"})),(0,t.kt)("center",null,(0,t.kt)("a",{href:"https://xkcd.com/1024"},"xkcd 1024: Error Code")),(0,t.kt)("admonition",{type:"note"},(0,t.kt)("p",{parentName:"admonition"},"Topik ini mengasumsikan program memberikan pesan error. Terdapat banyak kasus di mana hal ini tidak terjadi, seperti program yang dicompile dari bahasa C, yang hanya mengeluarkan pesan ",(0,t.kt)("inlineCode",{parentName:"p"},"Segmentation fault"),". Dalam kasus tersebut, teknik yang digunakan sudah berubah dengan melakukan debugging.")),(0,t.kt)("p",null,"Saat suatu error terjadi, program akan memberikan suatu pesan yang menjelaskan bahwa terjadi suatu error. Pesan yang diberikan dapat bervariasi, namun pada umumnya memiliki komponen berikut:"),(0,t.kt)("ol",null,(0,t.kt)("li",{parentName:"ol"},'Pesan error (Contoh: "An error has occured")'),(0,t.kt)("li",{parentName:"ol"},"Detail error"),(0,t.kt)("li",{parentName:"ol"},"Kode error")),(0,t.kt)("p",null,"Suatu program dapat saja tidak memiliki detail ataupun kode error, atau bahkan tidak memiliki keduanya. Tidak hanya itu, suatu program dapat saja tidak melakukan proteksi terhadap error, sehingga akan tertutup atau hang."),(0,t.kt)("p",null,'Pada umumnya, kamu ingin tahu detail suatu error, terutama dalam konteks pemroragaman. Sehingga kamu lebih baik memperhatikan kode dan detail dari error tersebut, dibandingkan pesan yang tidak memberikan kamu petunjuk. Sebagai contoh, pesan error yang berisi "Tidak dapat terkoneksi ke internet" lebih memberi banyak informasi daripada "Telah terjadi error". Oleh karena itu, ',(0,t.kt)("strong",{parentName:"p"},"pilah pesan-pesan yang muncul dan fokus terhadap pesan yang memberi kamu petunjuk terbaik"),"."),(0,t.kt)("p",null,"Berikut adalah contoh kode Python yang mencoba untuk meng-import modul yang tidak ada:"),(0,t.kt)("pre",null,(0,t.kt)("code",{parentName:"pre"},">>> import asdkjh\nTraceback (most recent call last):\n  File \"<stdin>\", line 1, in <module>\nModuleNotFoundError: No module named 'asdkjh'\n")),(0,t.kt)("p",null,"Dalam kasus ini, kamu cukup fokus terhadap pesan dan kode error yang diberikan, yaitu:"),(0,t.kt)("pre",null,(0,t.kt)("code",{parentName:"pre"},"ModuleNotFoundError: No module named 'asdkjh'\n")),(0,t.kt)("h2",{id:"membaca-stack-trace"},"Membaca Stack Trace"),(0,t.kt)("p",null,(0,t.kt)("em",{parentName:"p"},"atau traceback, backtrace, dan sejenisnya")),(0,t.kt)("p",null,"Jika kamu sedang melakukan pemrograman dan terjadi error, program biasanya akan memberi kamu stack trace. Stack trace adalah rekaman jejak dari fungsi-fungsi yang dipanggil hingga terjadi error. Contoh dari stack trace dapat dilihat sebagai berikut:"),(0,t.kt)("ul",null,(0,t.kt)("li",{parentName:"ul"},(0,t.kt)("p",{parentName:"li"},"Python"),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre",className:"language-py",metastring:'title="foobar.py" showLineNumbers',title:'"foobar.py"',showLineNumbers:!0},"def foo():\n    return bar()\n\ndef bar():\n    hello\n\nfoo()\n")),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre"},'Traceback (most recent call last):\n  File "foobar.py", line 7, in <module>\n    foo()\n  File "foobar.py", line 2, in foo\n    return bar()\n  File "foobar.py", line 5, in bar\n    hello\nNameError: name \'hello\' is not defined\n'))),(0,t.kt)("li",{parentName:"ul"},(0,t.kt)("p",{parentName:"li"},"Java"),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre",className:"language-java",metastring:'title="MainClass.java" showLineNumbers',title:'"MainClass.java"',showLineNumbers:!0},'public class MainClass {\n    public static void main(String args[]) {\n      foo();\n    }\n\n    public static void foo() {\n        bar();\n    }\n\n    public static void bar() {\n        throw new RuntimeException("Error here");\n    }\n}\n')),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre"},'Exception in thread "main" java.lang.RuntimeException: Error here\n  at MainClass.bar(MainClass.java:11)\n  at MainClass.foo(MainClass.java:7)\n  at MainClass.main(MainClass.java:3)\n')))),(0,t.kt)("p",null,"Tentu saja, stack trace dapat terlihat menyeramkan semakin besar program yang kalian buat. Namun, terdapat beberapa hal penting yang perlu kamu lakukan agar kamu dapat menentukan titik di mana program kamu error:"),(0,t.kt)("ol",null,(0,t.kt)("li",{parentName:"ol"},(0,t.kt)("p",{parentName:"li"},"Perhatikan alur penulisan dari stack trace.",(0,t.kt)("br",{parentName:"p"}),"\n","Sebagai contoh, di Python penulisan stack trace bersifat semakin ke bawah, semakin dekat dengan apa yang mengeluarkan error. Sebaliknya dengan java, semakin ke atas, semakin dekat dengan apa yang mengeluarkan error. Hal ini berguna untuk mempercepat menentukan titik di mana error itu terjadi."),(0,t.kt)("p",{parentName:"li"},"Tidak hanya itu, perlu diperhatikan bagaimana format penulisan dari stack trace, ini berguna agar kamu tahu hingga ke baris berapa yang menyebabkan error tersebut.")),(0,t.kt)("li",{parentName:"ol"},(0,t.kt)("p",{parentName:"li"},"Hanya perhatikan bagian yang relevan dengan kodemu.",(0,t.kt)("br",{parentName:"p"}),"\n","Secara umum, kamu hanya ingin melihat stack trace yang berasal dari kodemu. Artinya, kamu tidak perlu memerhatikan trace dari library, sehingga kamu dapat abaikan dan mendapat stack trace yang jauh lebih kecil."),(0,t.kt)("p",{parentName:"li"},"Sebagai contoh, kamu ingin membuat request ke suatu website menggunakan Python, dan kamu mendapat stack trace seperti berikut:"),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre"},'Traceback (most recent call last):\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 1348, in do_open\n    h.request(req.get_method(), req.selector, req.data, headers,\n  File "...\\3.10.4\\lib\\http\\client.py", line 1282, in request\n    self._send_request(method, url, body, headers, encode_chunked)\n  File "...\\3.10.4\\lib\\http\\client.py", line 1328, in _send_request\n    self.endheaders(body, encode_chunked=encode_chunked)\n  File "...\\3.10.4\\lib\\http\\client.py", line 1277, in endheaders\n    self._send_output(message_body, encode_chunked=encode_chunked)\n  File "...\\3.10.4\\lib\\http\\client.py", line 1037, in _send_output\n    self.send(msg)\n  File "...\\3.10.4\\lib\\http\\client.py", line 975, in send\n    self.connect()\n  File "...\\3.10.4\\lib\\http\\client.py", line 1447, in connect\n    super().connect()\n  File "...\\3.10.4\\lib\\http\\client.py", line 941, in connect\n    self.sock = self._create_connection(\n  File "...\\3.10.4\\lib\\socket.py", line 824, in create_connection\n    for res in getaddrinfo(host, port, 0, SOCK_STREAM):\n  File "...\\3.10.4\\lib\\socket.py", line 955, in getaddrinfo\n    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):\nsocket.gaierror: [Errno 11001] getaddrinfo failed\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File "foobar.py", line 2, in <module>\n    urllib.request.urlopen("https://askjdhieqf.com")\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 216, in urlopen\n    return opener.open(url, data, timeout)\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 519, in open\n    response = self._open(req, data)\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 536, in _open\n    result = self._call_chain(self.handle_open, protocol, protocol +\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 496, in _call_chain\n    result = func(*args)\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 1391, in https_open\n    return self.do_open(http.client.HTTPSConnection, req,\n  File "...\\3.10.4\\lib\\urllib\\request.py", line 1351, in do_open\n    raise URLError(err)\nurllib.error.URLError: <urlopen error [Errno 11001] getaddrinfo failed>\n')),(0,t.kt)("p",{parentName:"li"},"Tentu saja stack trace ini sangat membingungkan. Oleh karena itu, mari kita coba hapus baris-baris yang bukan merupakan bagian dari kode kita, namun tetap menyimpan detail error yang penting."),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre"},'Traceback (most recent call last):\n  ...\nsocket.gaierror: [Errno 11001] getaddrinfo failed\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File "foobar.py", line 2, in <module>\n    urllib.request.urlopen("https://askjdhieqf.com")\n  ...\nurllib.error.URLError: <urlopen error [Errno 11001] getaddrinfo failed>\n\n')),(0,t.kt)("p",{parentName:"li"},"Ternyata, file ",(0,t.kt)("inlineCode",{parentName:"p"},"foobar.py"),", baris kedua dengan isi ",(0,t.kt)("inlineCode",{parentName:"p"},'urllib.request.urlopen("https://askjdhieqf.com")')," menghasilkan error tersebut! Namun, kita tidak tahu mengapa error tersebut terjadi. Oleh karena itu, mari kita coba cari di Google mengenai pesan error (",(0,t.kt)("inlineCode",{parentName:"p"},"[Errno 11001] getaddrinfo failed"),") tersebut."),(0,t.kt)("p",{parentName:"li"},"Setelah melakukan pencarian, kita tahu bahwa ternyata error tersebut terjadi karena hostname tidak dapat ditemukan. Ini masuk akal sebab website ",(0,t.kt)("inlineCode",{parentName:"p"},"https://askjdhieqf.com")," tidak ada di internet.")),(0,t.kt)("li",{parentName:"ol"},(0,t.kt)("p",{parentName:"li"},"Perhatikan konteks dari kode yang error\nTerkadang, error tidak terjadi karena baris yang dicantumkan dalam stack trace, melainkan karena bagian kode lain yang menyebabkan baris tersebut error. Salah satu contohnya adalah penggunaan variable dengan tipe yang salah, atau nilai yang tidak benar."),(0,t.kt)("p",{parentName:"li"},"Sebagai contoh, perhatikan kode Java berikut:"),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre",className:"language-java",metastring:'title="Main.java" showLineNumbers',title:'"Main.java"',showLineNumbers:!0},'public class Main {\n    public static int sum(int[] nums, int i , int j) {\n        return nums[i] + nums[j];\n    }\n\n    public static void main(String args[]) {\n      int[] nums={ 1, 4, 2, 4, 5};\n      int i = 3;\n      int j = 9;\n      System.out.println("nums[i] + nums[j] = " + sum(nums, i, j));\n    }\n}\n')),(0,t.kt)("p",{parentName:"li"},"Berikut adalah stack trace dari program tersebut:"),(0,t.kt)("pre",{parentName:"li"},(0,t.kt)("code",{parentName:"pre"},'Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 9 out of bounds for length 5\n  at Main.sum(Main.java:3)\n  at Main.main(Main.java:10)\n')),(0,t.kt)("p",{parentName:"li"},'Jika kita hanya mengikuti stack trace, kita tidak dapat melihat keanehan apapun karena operasi tersebut "benar". Namun, jika kita melihat konteks error serta kode di sekitar, kita dapat tahu bahwa penyebab error ini berada pada variable ',(0,t.kt)("inlineCode",{parentName:"p"},"j")," yang bernilai 9, sehingga program mencoba untuk mengambil indeks ke-9. Hal ini tidak benar mengingat array ",(0,t.kt)("inlineCode",{parentName:"p"},"nums")," hanya memiliki 6 elemen."))),(0,t.kt)("admonition",{type:"info"},(0,t.kt)("p",{parentName:"admonition"},"Untuk bagaimana cara mencari solusi error, kamu dapat kembali ke topik ",(0,t.kt)("a",{parentName:"p",href:"/gsgs#mencari-solusi-error"},"Mesin Pencari"),".")))}k.isMDXComponent=!0}}]);